File: runtime/scala/src/main/scala/dev/argon/verilization/scala_runtime/VLQ.scala
package dev.argon.verilization.scala_runtime

import zio.{ZIO, IO, Chunk}
import java.math.BigInteger
import scala.math.BigInt

object VLQ {


    private final case class OutputState(currentByte: Byte, outBitIndex: Int)

    def encodeVLQ[R, E](writer: FormatWriter[R, E], isSigned: Boolean, n: BigInt): ZIO[R, E, Unit] = {
        val nBytes = (if(isSigned && n < 0) n + 1 else n).abs.toByteArray

        def putBit(b: Boolean, state: OutputState): ZIO[R, E, OutputState] =
            if(state.outBitIndex > 6) // Only use 7 bits, 8th bit is for tag to indicate more data
                writer.writeByte((state.currentByte | 0x80).toByte) *> putBit(b, OutputState(0, 0))
            else
                IO.succeed(OutputState(
                    currentByte = if(b) (state.currentByte | (1 << state.outBitIndex)).toByte else state.currentByte,
                    outBitIndex = state.outBitIndex + 1,
                ))

        def putSign(sign: Boolean, state: OutputState): ZIO[R, E, OutputState] =
            if(state.outBitIndex != 6)
                // Pad out until the sign bit
                putBit(false, state).flatMap { state =>
                    putSign(sign, state)
                }
            else
                putBit(sign, state)
            
        def finish(state: OutputState): ZIO[R, E, Unit] =
            writer.writeByte(state.currentByte)

        def iterBits(byteIndex: Int, bitIndex: Int, zeroCount: Int, outputState: OutputState): ZIO[R, E, Unit] =
            if(byteIndex < 0) {
                if(isSigned) putSign(n < 0, outputState).flatMap(finish)
                else finish(outputState)
            }
            else {
                val bit = (nBytes(byteIndex) & (1 << bitIndex)) != 0

                val (byteIndex2, bitIndex2) =
                    if(bitIndex + 1 > 7)
                        (byteIndex - 1, 0)
                    else
                        (byteIndex, bitIndex + 1)

                def putZeroes(zeroCount: Int, outputState: OutputState): ZIO[R, E, OutputState] =
                    if(zeroCount > 0) putBit(false, outputState).flatMap { outputState => putZeroes(zeroCount - 1, outputState) }
                    else IO.succeed(outputState)
                    
                if(bit)
                    putZeroes(
                        zeroCount = zeroCount,
                        outputState = outputState,
                    ).flatMap { outputState =>
                        putBit(true, outputState)
                    }.flatMap { outputState =>
                        iterBits(
                            byteIndex = byteIndex2,
                            bitIndex = bitIndex2,
                            zeroCount = 0,
                            outputState = outputState,
                        )
                    }
                else
                    iterBits(
                        byteIndex = byteIndex2,
                        bitIndex = bitIndex2,
                        zeroCount = zeroCount + 1,
                        outputState = outputState,
                    )
            }

        iterBits(
            byteIndex = nBytes.length - 1,
            bitIndex = 0,
            zeroCount = 0,
            outputState = OutputState(currentByte = 0, outBitIndex = 0)
        )
    }


    private final case class BigIntBuildState(currentByte: Byte, otherBytes: Chunk[Byte], bitIndex: Int) {

        def putBit(b: Boolean): BigIntBuildState = {
            val newByte = if(b) (currentByte | (1 << bitIndex)).toByte else currentByte

            if(bitIndex + 1 > 7)
                BigIntBuildState(
                    currentByte = 0,
                    otherBytes = newByte +: otherBytes,
                    bitIndex = 0, 
                )
            else
                BigIntBuildState(
                    currentByte = newByte,
                    otherBytes = otherBytes,
                    bitIndex = bitIndex + 1,
                )
        }

        def toByteArray: Array[Byte] =
            (currentByte +: otherBytes).toArray

    }

    def decodeVLQ[R, E](reader: FormatReader[R, E], isSigned: Boolean): ZIO[R, E, BigInt] = {

        def processBits(state: BigIntBuildState, b: Byte, i: Int, n: Int): BigIntBuildState =
            if(i < n) processBits(state.putBit((b & (1 << i)) != 0), b, i + 1, n)
            else state

        def readBytes(state: BigIntBuildState): ZIO[R, E, BigInt] =
            reader.readByte().flatMap { b =>
                if((b & 0x80) != 0)
                    readBytes(processBits(state, b, 0, 7))
                else {
                    val signbit = isSigned && (b & 0x40) != 0
                    val sign = if(signbit) -1 else 1
                    val bigInteger = new BigInteger(sign, processBits(state, b, 0, if(isSigned) 6 else 7).toByteArray)
                    IO.succeed(BigInt(bigInteger) - (if(signbit) 1 else 0))
                }
            }

        readBytes(BigIntBuildState(0, Chunk.empty, 0))
    }



}

File: runtime/java/src/main/java/dev/argon/verilization/runtime/VLQ.java
package dev.argon.verilization.runtime;

import java.io.IOException;
import java.math.BigInteger;

abstract class VLQ {
	private VLQ() {}

	private static final class Encoder {
		public Encoder(FormatWriter writer) {
			this.writer = writer;
		}

		private final FormatWriter writer;
		private int outBitIndex = 0;
		private byte currentByte = 0;

		public void putBit(boolean b) throws IOException {
			if(outBitIndex > 6) { // Only use 7 bits, 8th bit is for tag to indicate more data
				writer.writeByte((byte)(currentByte | 0x80));
				outBitIndex = 0;
				currentByte = 0;
			}

			if(b) currentByte |= 1 << outBitIndex;
			outBitIndex += 1;
		}
		
		public void putSign(boolean sign) throws IOException {
			while(outBitIndex != 6) { // Pad out until the sign bit
				putBit(false);
			}

			putBit(sign);
		}

		public void finish() throws IOException {
			writer.writeByte(currentByte);
		}
	}

	public static void encodeVLQ(FormatWriter writer, boolean isSigned, BigInteger n) throws IOException {
		var nBytes = ((isSigned && n.signum() < 0) ? n.add(BigInteger.ONE) : n).abs().toByteArray();

		var encoder = new Encoder(writer);
		int byteIndex = nBytes.length - 1;
		int bitIndex = 0;
		int zeroCount = 0;

		while(byteIndex >= 0) {
			boolean bit = (nBytes[byteIndex] & (1 << bitIndex)) != 0;
			if(bit) {
				for(int i = 0; i < zeroCount; ++i) {
					encoder.putBit(false);
				}
				zeroCount = 0;
				encoder.putBit(true);
			}
			else {
				++zeroCount;
			}

			++bitIndex;
			if(bitIndex > 7) {
				bitIndex = 0;
				--byteIndex;
			}
		}

		if(isSigned) {
			encoder.putSign(n.signum() < 0);
		}

		encoder.finish();
	}

	private static class BigIntegerBuilder {

		private static final int INITIAL_SIZE = 8;
		private byte[] data = new byte[INITIAL_SIZE];
		private int byteIndex = INITIAL_SIZE - 1;
		private int bitIndex = 0;

		public void putBit(boolean b) {
			if(b) data[byteIndex] |= 1 << bitIndex;
			++bitIndex;
			if(bitIndex > 7) {
				bitIndex = 0;
				--byteIndex;

				if(byteIndex < 0) {
					byte[] newData = new byte[data.length * 2];
					for(int i = 0; i < data.length; ++i) {
						newData[data.length + i] = data[i];
					}

					byteIndex += data.length;
				}
			}
		}

		public BigInteger bigInteger(boolean sign) {
			var result = new BigInteger(sign ? -1 : 1, data);
			if(sign) {
				return result.subtract(BigInteger.ONE);
			}
			else {
				return result;
			}
		}
	}

	public static BigInteger decodeVLQ(FormatReader reader, boolean isSigned) throws IOException {

		var builder = new BigIntegerBuilder();
		
		byte b = reader.readByte();
		while((b & 0x80) != 0) {
			for(int i = 0; i < 7; ++i) {
				builder.putBit((b & (1 << i)) != 0);
			}
			b = reader.readByte();
		}


		for(int i = 0; i < (isSigned ? 6 : 7); ++i) {
			builder.putBit((b & (1 << i)) != 0);
		}

		boolean sign = isSigned && (b & 0x40) != 0;

		return builder.bigInteger(sign);
	}
}
